File: rust_code/pi_afr/src/bin/pushButton_led.rs
use rppal::gpio::{Gpio, Level, Trigger};
use std::error::Error;
use std::sync::{Arc, atomic::{AtomicBool, Ordering}};
use std::time::Duration;

fn main() -> Result<(), Box<dyn Error>> {
    // BCM pin 21 = Physical pin 40 (Button)
    // BCM pin 20 = Physical pin 38 (LED)
    let button_pin_bcm = 21;
    let led_pin_bcm = 20;
    
    // Initialize GPIO
    let gpio = Gpio::new()?;
    
    // Set up the button pin with internal pull-up resistor
    let mut button = gpio.get(button_pin_bcm)?.into_input_pullup();
    
    // Set up the LED pin as output
    let mut led = gpio.get(led_pin_bcm)?.into_output();
    led.set_low(); // Start with LED off
    
    // Set up interrupt handling
    let running = Arc::new(AtomicBool::new(true));
    let r = running.clone();
    
    // Handle Ctrl+C
    ctrlc::set_handler(move || {
        r.store(false, Ordering::SeqCst);
        println!("\nReceived Ctrl+C, exiting...");
    })?;
    
    // Display initial state
    println!("Push button monitoring started. Press Ctrl+C to exit.");
    println!("Current state: {} ({})",
             if button.is_high() { "HIGH" } else { "LOW" },
             if button.is_high() { "Button NOT pressed" } else { "Button pressed" });
    
    // Initial LED state based on button state
    if button.is_high() { led.set_low(); } else { led.set_high(); }
    
    // Set up interrupt for both rising and falling edge
    button.set_interrupt(Trigger::Both)?;
    
    while running.load(Ordering::SeqCst) {
        match button.poll_interrupt(true, Some(Duration::from_millis(100)))? {
            Some(level) => {
                match level {
                    Level::Low => {  // Button pressed
                        println!("Button pressed! Turning LED ON");
                        led.set_high();
                    },
                    Level::High => {  // Button released
                        println!("Button released! Turning LED OFF");
                        led.set_low();
                    },
                }
            },
            None => {} // Timeout, do nothing
        }
    }
    
    // Turn off LED before exiting
    led.set_low();
    println!("Program completed");
    
    Ok(())
}

File: python_code/bin/pushButton_led.py
import RPi.GPIO as GPIO
import time

# Setup
GPIO.setwarnings(False)
GPIO.setmode(GPIO.BOARD)

# Define pins
BUTTON_PIN = 40  # Your button is on pin 40
LED_PIN = 38  # LED positive leg connected to pin 38

# Setup button pin with internal pull-up resistor
GPIO.setup(BUTTON_PIN, GPIO.IN, pull_up_down=GPIO.PUD_UP)

# Setup LED pin as output
GPIO.setup(LED_PIN, GPIO.OUT)
GPIO.output(LED_PIN, GPIO.LOW)  # Start with LED off

def button_callback(channel):
    if GPIO.input(BUTTON_PIN) == GPIO.LOW:  # Button is pressed
        print("Button pressed! Turning LED ON")
        GPIO.output(LED_PIN, GPIO.HIGH)
    else:  # Button is released
        print("Button released! Turning LED OFF")
        GPIO.output(LED_PIN, GPIO.LOW)

# Register interrupt for both rising and falling edge
GPIO.add_event_detect(BUTTON_PIN, GPIO.BOTH, callback=button_callback, bouncetime=200)

try:
    print("Push button monitoring started. Press Ctrl+C to exit.")
    print("Current state: {} ({})".format(
        "HIGH" if GPIO.input(BUTTON_PIN) == GPIO.HIGH else "LOW",
        "Button NOT pressed" if GPIO.input(BUTTON_PIN) == GPIO.HIGH else "Button pressed"
    ))
    
    # Initial LED state based on button state
    GPIO.output(LED_PIN, GPIO.LOW if GPIO.input(BUTTON_PIN) == GPIO.HIGH else GPIO.HIGH)
    
    while True:
        time.sleep(0.1)  # Just keep the program running

except KeyboardInterrupt:
    print("\nProgram stopped by user")
finally:
    # Turn off LED and clean up
    GPIO.output(LED_PIN, GPIO.LOW)
    GPIO.cleanup()
    print("GPIO cleanup completed")
